File: app/local_plugins/stt_azure/android/src/main/kotlin/com/aiagent/stt_azure/SttAzurePlugin.kt
package com.aiagent.stt_azure

import android.util.Log
import com.aiagent.plugin_interface.NativeServiceRegistry
import com.microsoft.cognitiveservices.speech.*
import com.microsoft.cognitiveservices.speech.audio.AudioConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*

/**
 * SttAzurePlugin — Azure 语音识别 Flutter 插件
 *
 * 双重角色：
 * 1. NativeServiceRegistry 注册：Agent 类型插件通过 NativeSttService 接口直接调用
 * 2. MethodChannel/EventChannel：保留 Dart 桥接（向后兼容）
 */
class SttAzurePlugin : FlutterPlugin {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    private val mainScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var recognizer: SpeechRecognizer? = null
    private var speechConfig: SpeechConfig? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val context = binding.applicationContext

        // ── 注册到 NativeServiceRegistry（供 Agent 类型插件原生调用）──
        NativeServiceRegistry.registerStt("azure") { SttAzureService(context) }
        Log.d("SttAzurePlugin", "Registered NativeSttService vendor=azure")

        // ── 保留 MethodChannel/EventChannel（Dart 桥接，向后兼容）──
        methodChannel = MethodChannel(binding.binaryMessenger, "stt_azure/commands")
        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "initialize" -> {
                    val key = call.argument<String>("apiKey")!!
                    val region = call.argument<String>("region")!!
                    val lang = call.argument<String>("language") ?: "zh-CN"
                    mainScope.launch(Dispatchers.IO) {
                        try {
                            initialize(key, region, lang)
                            withContext(Dispatchers.Main) { result.success(null) }
                        } catch (e: Exception) {
                            withContext(Dispatchers.Main) { result.error("STT_INIT", e.message, null) }
                        }
                    }
                }
                "startListening" -> {
                    startListening()
                    result.success(null)
                }
                "stopListening" -> {
                    stopListening()
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        }

        eventChannel = EventChannel(binding.binaryMessenger, "stt_azure/events")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(args: Any?, sink: EventChannel.EventSink?) { eventSink = sink }
            override fun onCancel(args: Any?) { eventSink = null }
        })
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        recognizer?.close()
        speechConfig?.close()
        mainScope.cancel()
    }

    // ─────────────────────────────────────────────────
    // Dart MethodChannel 实现（向后兼容）
    // ─────────────────────────────────────────────────

    private fun initialize(apiKey: String, region: String, language: String) {
        recognizer?.close()
        speechConfig?.close()

        speechConfig = SpeechConfig.fromSubscription(apiKey, region).apply {
            speechRecognitionLanguage = language
            setProperty(PropertyId.SpeechServiceResponse_RequestDetailedResultTrueFalse, "true")
        }

        val audioConfig = AudioConfig.fromDefaultMicrophoneInput()
        recognizer = SpeechRecognizer(speechConfig, audioConfig).apply {
            recognizing.addEventListener { _, e ->
                pushEvent(mapOf("kind" to "partialResult", "text" to e.result.text))
            }
            recognized.addEventListener { _, e ->
                if (e.result.reason == ResultReason.RecognizedSpeech) {
                    pushEvent(mapOf("kind" to "finalResult", "text" to e.result.text))
                }
            }
            sessionStarted.addEventListener { _, _ ->
                pushEvent(mapOf("kind" to "listeningStarted"))
            }
            sessionStopped.addEventListener { _, _ ->
                pushEvent(mapOf("kind" to "listeningStopped"))
            }
            speechStartDetected.addEventListener { _, _ ->
                pushEvent(mapOf("kind" to "vadSpeechStart"))
            }
            speechEndDetected.addEventListener { _, _ ->
                pushEvent(mapOf("kind" to "vadSpeechEnd"))
            }
            canceled.addEventListener { _, e ->
                if (e.reason == CancellationReason.Error) {
                    pushEvent(mapOf(
                        "kind" to "error",
                        "errorCode" to e.errorCode.toString(),
                        "errorMessage" to e.errorDetails,
                    ))
                }
            }
        }
    }

    private fun startListening() {
        val r = recognizer ?: return
        mainScope.launch(Dispatchers.IO) {
            r.startContinuousRecognitionAsync().get()
        }
    }

    private fun stopListening() {
        val r = recognizer ?: return
        mainScope.launch(Dispatchers.IO) {
            r.stopContinuousRecognitionAsync().get()
        }
    }

    private fun pushEvent(data: Map<String, Any?>) {
        mainScope.launch { eventSink?.success(data) }
    }
}

File: app/local_plugins/stt_azure/ios/Classes/SttAzurePlugin.swift
import Flutter
import MicrosoftCognitiveServicesSpeech

/**
 * SttAzurePlugin — Azure 语音识别 iOS 实现
 *
 * 推送 7 种 STT 事件（与 Android 完全对称）：
 *   listeningStarted / vadSpeechStart / vadSpeechEnd /
 *   partialResult / finalResult / listeningStopped / error
 */
public class SttAzurePlugin: NSObject, FlutterPlugin {

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?

    private var recognizer: SPXSpeechRecognizer?
    private var speechConfig: SPXSpeechConfiguration?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = SttAzurePlugin()

        instance.methodChannel = FlutterMethodChannel(
            name: "stt_azure/commands",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: instance.methodChannel!)

        instance.eventChannel = FlutterEventChannel(
            name: "stt_azure/events",
            binaryMessenger: registrar.messenger()
        )
        instance.eventChannel?.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "initialize":
            let key = args["apiKey"] as! String
            let region = args["region"] as! String
            let lang = args["language"] as? String ?? "zh-CN"
            initialize(apiKey: key, region: region, language: lang)
            result(nil)
        case "startListening":
            startListening()
            result(nil)
        case "stopListening":
            stopListening()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // ─────────────────────────────────────────────────
    // MARK: — 实现
    // ─────────────────────────────────────────────────

    private func initialize(apiKey: String, region: String, language: String) {
        recognizer = nil
        speechConfig = nil

        guard let config = try? SPXSpeechConfiguration(subscription: apiKey, region: region) else {
            pushEvent(["kind": "error", "errorCode": "init_failed", "errorMessage": "SPXSpeechConfiguration init error"])
            return
        }
        config.speechRecognitionLanguage = language
        speechConfig = config

        let audioConfig = SPXAudioConfiguration()
        guard let rec = try? SPXSpeechRecognizer(speechConfiguration: config, audioConfiguration: audioConfig) else {
            pushEvent(["kind": "error", "errorCode": "recognizer_init_failed"])
            return
        }

        // Partial result
        rec.addRecognizingEventHandler { [weak self] (_: SPXSpeechRecognizer, e: SPXSpeechRecognitionEventArgs) in
            self?.pushEvent(["kind": "partialResult", "text": e.result.text ?? ""])
        }
        // Final result
        rec.addRecognizedEventHandler { [weak self] (_: SPXSpeechRecognizer, e: SPXSpeechRecognitionEventArgs) in
            if e.result.reason == SPXResultReason.recognizedSpeech {
                self?.pushEvent(["kind": "finalResult", "text": e.result.text ?? ""])
            }
        }
        // Session start
        rec.addSessionStartedEventHandler { [weak self] (_: SPXRecognizer, _: SPXSessionEventArgs) in
            self?.pushEvent(["kind": "listeningStarted"])
        }
        // Session stop
        rec.addSessionStoppedEventHandler { [weak self] (_: SPXRecognizer, _: SPXSessionEventArgs) in
            self?.pushEvent(["kind": "listeningStopped"])
        }
        // VAD speech start
        rec.addSpeechStartDetectedEventHandler { [weak self] (_: SPXRecognizer, _: SPXRecognitionEventArgs) in
            self?.pushEvent(["kind": "vadSpeechStart"])
        }
        // VAD speech end
        rec.addSpeechEndDetectedEventHandler { [weak self] (_: SPXRecognizer, _: SPXRecognitionEventArgs) in
            self?.pushEvent(["kind": "vadSpeechEnd"])
        }
        // Canceled / error
        rec.addCanceledEventHandler { [weak self] (_: SPXRecognizer, e: SPXSpeechRecognitionCanceledEventArgs) in
            if e.reason == SPXCancellationReason.error {
                self?.pushEvent([
                    "kind": "error",
                    "errorCode": "\(e.errorCode)",
                    "errorMessage": e.errorDetails ?? ""
                ])
            }
        }

        recognizer = rec
    }

    private func startListening() {
        guard let rec = recognizer else { return }
        DispatchQueue.global(qos: .userInitiated).async {
            try? rec.startContinuousRecognition()
        }
    }

    private func stopListening() {
        guard let rec = recognizer else { return }
        DispatchQueue.global(qos: .userInitiated).async {
            try? rec.stopContinuousRecognition()
        }
    }

    private func pushEvent(_ data: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(data)
        }
    }
}

// ─────────────────────────────────────────────────
// MARK: — FlutterStreamHandler
// ─────────────────────────────────────────────────

extension SttAzurePlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        recognizer = nil
        return nil
    }
}
